src/libsyntax/ext/base.rs

```diff
@@ -474,6 +474,35 @@ impl MacResult for DummyResult {
 pub type BuiltinDeriveFn =
     for<'cx> fn(&'cx mut ExtCtxt, Span, &MetaItem, &Annotatable, &mut FnMut(Annotatable));
 
+/// Represents different kinds of macro invocations that can be resolved.
+#[derive(Clone, Copy)]
+pub enum MacroInvocKind {
+    /// A bang macro - foo!()
+    Bang,
+    /// An attribute macro - #[foo]
+    Attr,
+    /// A derive attribute macro - #[derive(Foo)]
+    Derive,
+}
+
+impl MacroInvocKind {
+    /// Returns true if the syntax extension is allowed for this macro invocation.
+    pub fn allow_ext(&self, _ext: &Rc<SyntaxExtension>) -> bool {
```

Comment:
I didn't think that there was a 1-1 mapping, but I can see that there seems to be.